scipy/special/_lambertw.h

```diff
@@ -0,0 +1,167 @@
+/* Implementation of the Lambert W function [1]. Based on MPMath
+ *  Implementation [2], and documentation [3].
+ *
+ * Copyright: Yosef Meller, 2009
+ * Authoer email: mellerf@netvision.net.il
+ *
+ * Distributed under the same license as SciPy
+ * Translated into C++ by SciPy developers, 2023.
+ *
+ * References:
+ * [1] On the Lambert W function, Adv. Comp. Math. 5 (1996) 329-359,
+ *     available online: http://www.apmaths.uwo.ca/~djeffrey/Offprints/W-adv-cm.pdf
+ * [2] mpmath source code, Subversion revision 990
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/functions.py?spec=svn994&r=992
+ * [3] mpmath source code, Subversion revision 994
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/function_docs.py?spec=svn994&r=994
+
+ * TODO: use a series expansion when extremely close to the branch point
+ * at `-1/e` and make sure that the proper branch is chosen there.
+ */
+
+
+#ifndef LAMBERTW_H
+#define LAMBERTW_H
+
+#include <cmath>
+#include <complex>
+#include <numpy/npy_math.h>
```

Comment:
The `#include <numpy/npy_math.h>` is problematic for other array libraries. It may not exist and it certainly won't work with NVCC's JIT. What I'd suggest is we create a new single header file, just for SciPy, that brings together all the special function C++ implementations and any NumPy stuff they need.

Reply:
Thanks @izaid, definitely an oversight on my part. That's exactly how we should do it.